enterprise/app/code/code.tsx

```diff
@@ -0,0 +1,383 @@
+import React from "react";
+import rpcService from "../../../app/service/rpc_service";
+import { User } from "../../../app/auth/auth_service";
+import SidebarNodeComponent from "./code_sidebar_node";
+import { Subscription } from "rxjs";
+import * as monaco from "monaco-editor";
+import { Octokit } from "octokit";
+import DiffMatchPatch from "diff-match-patch";
+import { createPullRequest } from "octokit-plugin-create-pull-request";
+
+const MyOctokit = Octokit.plugin(createPullRequest);
+
+interface Props {
+  user: User;
+  hash: string;
+  path: string;
+  search: URLSearchParams;
+}
+interface State {
+  owner: string;
+  repo: string;
+  repoResponse: any;
+  treeShaToExpanded: Map<string, boolean>;
+  treeShaToChildrenMap: Map<string, any[]>;
+  treeShaToPathMap: Map<string, string>;
+  fullPathToModelMap: Map<string, any>;
+  originalFileContents: string;
+  currentFilePath: string;
+  changes: Map<string, string>;
+  pathToIncludeChanges: Map<string, boolean>;
+
+  requestingReview: boolean;
+}
+
+const dmp = new DiffMatchPatch.diff_match_patch();
+
+// TODO(siggisim): Implement build and test
+// TODO(siggisim): Add links to the code editor from anywhere we reference a repo
+// TODO(siggisim): Add branch / workspace selection
+// TODO(siggisim): Add currently selected file name
+// TODO(siggisim): Add some form of search
+export default class CodeComponent extends React.Component<Props> {
+  props: Props;
+
+  state: State = {
+    owner: "",
+    repo: "",
+    repoResponse: undefined,
+    treeShaToExpanded: new Map<string, boolean>(),
+    treeShaToChildrenMap: new Map<string, any[]>(),
+    treeShaToPathMap: new Map<string, string>(),
```

Comment:
I was going to ask if it would make sense to store this state locally at each node instead of having a global flat map like this -- but I guess the advantage to doing it this way is so that you can cache the child node contents and avoid having to re-make the same requests even if a directory is collapsed then expanded again? (If I got that right, might be worth adding a little comment explaining the advantage, so someone like me doesn't try to turn it into a tree structure :stuck_out_tongue: )

Reply:
I'm generally against most forms of local state - the one of these I could see moving to local component state is whether or not the child node is expanded.

Often some state like this initially starts as being useful for one child component - and then later you realize that some other component (say a breadcrumb dropdown folder picker) needs the state and it becomes a paint to figure out some way to share this state.